### ISSUE TITLE
Rule Change: no-param-reassign, option to ignore reassignment at the top of the function

### ISSUE THREAD

**Issue by ilyub** [eslint/eslint#15628]
### What rule do you want to change?

no-param-reassign

### What change to do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new option

### Example code

```js
export function getMainOption(
  config: string | { readonly mainOption: string }
): string {
  config = typeof config === "string" ? { mainOption: config } : config;

  return config.mainOption;
}
```

### What does the rule currently do for this code?

Shows warning for "config =".

### What will the rule do after it's changed?

It will ignore this reassignment with "ignoreReassignmentAtFunctionTop" option.

### Participation

- [ ] I am willing to submit a pull request to implement this change.

### Additional comments

It is common situation that parameter can be passed in several aternative forms.
In my sample you can pass "options" parameter as an object or you can pass only main option as a string.
It is often desirable to reduce this parameter to a single form at the top of the function, but it is inconvenient to use new name (e.g. "reducedConfig" or "normalizedConfig").

Reassignment at the top of the function is transparent enough.
For me, adding new variable (e.g. "reducedConfig" or "normalizedConfig") is worth.

I would still use no-param-reassign rule, but with ignoreReassignmentAtFunctionTop option.

**Comment by nzakas** [comment]
Thanks for the input. As [stated in the README](https://github.com/eslint/eslint#stylistic-rule-updates), we actually aren't making any further updates to rules that enforce stylistic preferences. If you have a strong preference for such a change, you can always copy the existing rule into a custom rule and make it behave the way you want.
